triage/summarize/map_abstractions.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file contains abstractions of certain complex map types that are used often throughout
+summarize.go. They also have some often-used methods such as getting just the keys, or getting the
+key-value pairs as a map.
+*/
+
+package summarize
+
+import (
+	"sort"
+)
+
+// map[string][]failure type alias
+
+// failuresGroup maps strings to failure slices.
+type failuresGroup map[string][]failure
+
+// failuresGroupPair is a representation of a failuresGroup key-value mapping as a two-element
+// struct.
+type failuresGroupPair struct {
+	Key      string    `json:"key"`
+	Failures []failure `json:"failures"`
+}
+
+// keys provides the failuresGroup's keys as a string slice.
+func (fg *failuresGroup) keys() []string {
+	result := make([]string, len(*fg))
+
+	iter := 0
+	for key := range *fg {
+		result[iter] = key
+		iter++
+	}
+
+	return result
+}
+
+// asSlice returns the failuresGroup as a failuresGroupPair slice.
+func (fg *failuresGroup) asSlice() []failuresGroupPair {
+	result := make([]failuresGroupPair, len(*fg))
+
+	iter := 0
+	for str, failures := range *fg {
+		result[iter] = failuresGroupPair{str, failures}
+		iter++
+	}
+
+	return result
+}
+
+// sortByNumberOfFailures returns a failuresGroupPair slice sorted by the number of failures in each
+// pair, descending. If the number of failures is the same for two pairs, they are sorted alphabetically
+// by their keys.
+func (fg *failuresGroup) sortByNumberOfFailures() []failuresGroupPair {
```

Comment:
maybe `sortByGreatestFailures` so it is obvious and you can remove the comment below

Reply:
Addressed in #18882

Reply:
Addressed in #18882